net.go

```diff
@@ -118,6 +120,233 @@ func (e *EurekaConnection) GetApps() (map[string]*Application, error) {
+	}
+	return count
+}
+
+type instanceQueryOptions struct {
+	// secure indicates whether to query a secure or an insecure VIP address.
+	secure bool
+	// predicate guides filtering, indicating whether to retain an instance when it returns true or
+	// drop it when it returns false.
+	predicate func(*Instance) bool
+	// intn behaves like the rand.Rand.Intn function, aiding in randomizing the order of the result
+	// sequence when non-nil.
+	intn func(int) int
+}
+
+// InstanceQueryOption is a customization supplied to instance query functions like
+// GetInstancesByVIPAddress to tailor the set of instances returned.
+type InstanceQueryOption func(*instanceQueryOptions) error
+
+func retainIfStatusIs(status StatusType, o *instanceQueryOptions) {
+	if prev := o.predicate; prev != nil {
+		o.predicate = func(instance *Instance) bool {
+			return prev(instance) || instance.Status == status
+		}
+	} else {
+		o.predicate = func(instance *Instance) bool {
+			return instance.Status == status
+		}
+	}
+}
+
+// WithStatus restricts the set of instances returned to only those with the given status.
+//
+// Supplying multiple options produced by this function applies their logical disjunction.
+func WithStatus(status StatusType) InstanceQueryOption {
+	return func(o *instanceQueryOptions) error {
+		if len(status) == 0 {
+			return errors.New("invalid instance status")
+		}
+		retainIfStatusIs(status, o)
+		return nil
+	}
+}
+
+func makeUncheckedStatusMatchingOption(status StatusType) InstanceQueryOption {
+	return func(o *instanceQueryOptions) error {
+		retainIfStatusIs(status, o)
+		return nil
+	}
+}
+
+// ThatAreUp restricts the set of instances returned to only those with status UP.
+//
+// Combining the options produced by this function with those produced by WithStatus applies their
+// logical disjunction.
+func ThatAreUp() InstanceQueryOption {
+	return makeUncheckedStatusMatchingOption(UP)
+}
+
+// Shuffled requests randomizing the order of the sequence of instances returned, using the default
+// shared rand.Source.
+func Shuffled() InstanceQueryOption {
+	return func(o *instanceQueryOptions) error {
+		o.intn = rand.Intn
+		return nil
+	}
+}
+
+// ShuffledWith requests randomizing the order of the sequence of instances returned, using the
+// supplied source of random numbers.
+func ShuffledWith(r *rand.Rand) InstanceQueryOption {
+	return func(o *instanceQueryOptions) error {
+		o.intn = r.Intn
+		return nil
+	}
+}
+
+func shuffleInstances(instances []*Instance, intn func(int) int) {
+	count := len(instances)
+	if count < 2 {
+		return
+	}
+	if intn(2) == 0 {
+		instances[1], instances[0] = instances[0], instances[1]
+	}
+	for i := 2; i != count; i++ {
+		if j := intn(i + 1); j != i {
+			instances[i], instances[j] = instances[j], instances[i]
+		}
+	}
+}
+
+func filterInstances(instances []*Instance, pred func(*Instance) bool) []*Instance {
+	for firstBegin, instance := range instances {
+		if !pred(instance) {
+			continue
+		}
```

Comment:
I wrote the simple version first, stared at it, and thought, "Well, that's dumb. Why are we copying all the elements into a fresh sequence when we can take a slice?" We can't _always_ take a slice, of course, but we can when there is only one (contiguous) sequence of items to retain.

It's for performance reasons, yes, and I'd hoist this function out to a more general library if either Go had templates or generics or if I could see a way to make it work on more types than `*fargo.Instance`. Perhaps with reflection, though that may be enough to overwhelm the benefit of avoiding allocation and copying. It was enjoyable to write, and I'd like to be able to use it in more situations.

The greatest number of instances I've seen registered with a single VIP address is around 300. If I can over the next day or so, I'll write a benchmark to examine whether this implementation makes any defensible difference. Let me get back to you on it.